R/conformal_infer.R

```diff
@@ -0,0 +1,455 @@
+  # purrr::map_dfr(
+  furrr::future_map_dfr(
+    new_data,
+    optimize_one,
+    model = model,
+    train_data = train_data,
+    level = level,
+    ctrl,
+    .progress = ctrl$progress,
+    .options = furrr::furrr_options(seed = ctrl$seed, stdout = FALSE)
+  )
+}
+
+get_diff <- function(val, ...) {
+  res <- trial_fit(val, ...)
+  res$difference
+}
+
+optimize_one <- function(new_data, model, train_data, level, ctrl) {
+  y_name <- get_outcome_name(model)
+
+  pkg_inst <- lapply(ctrl$required_pkgs, rlang::is_installed)
+  pred_val <- new_data$.pred
+  bound <- new_data$.bound
+  new_data <- dplyr::select(new_data, -.pred, -.bound)
+  new_data[[y_name]] <- NA_real_
+  trial_data <- dplyr::bind_rows(train_data, new_data)
+
+  upper <-
+    try(
+      stats::uniroot(get_diff, c(pred_val, pred_val + bound),
+                     maxiter = ctrl$max_iter, tol = ctrl$tolerance,
+                     extendInt = "upX",
+                     trial_data, model, level),
+      silent = TRUE)
+
+  lower <-
+    try(
+      stats::uniroot(get_diff, c(pred_val - bound, pred_val),
+                     maxiter = ctrl$max_iter, tol = ctrl$tolerance,
+                     extendInt = "downX",
+                     trial_data, model, level),
+      silent = TRUE)
+
+  dplyr::tibble(
+    .pred_lower = get_root(lower, ctrl),
+    .pred_upper = get_root(upper, ctrl)
+  )
+}
+
+get_root <- function(x, ctrl) {
+  if (inherits(x, "try-error")) {
+    rlang::warn("Could not finish the search process due to errors.")
+    return(NA_real_)
+  }
+  if (x$iter == ctrl$max_iter) {
+    rlang::warn("Search did not converge.")
+    return(NA_real_)
+  }
+  return(x$root)
+}
+
+# ------------------------------------------------------------------------------
+
+#' Controlling the numeric details for conformal inference
+#'
+#'
+#' @param method The method for computing the intervals. The options are
+#' `'search'` (using) [stats::uniroot()], and `'grid'`.
+#' @param trial_points When `method = "grid"`, how many points should be
+#' evaluated?
+#' @param var_multiplier A multiplier for the variance model that determines the
+#' possible range of the bounds.
+#' @param max_iter When `method = "search"`, the maximum number of iterations.
+#' @param tolerance Tolerance value passed to [all.equal()] to determine
+#' convergence during the search computations.
+#' @param progress Should a progress bar be used to track execution?
+#' @param required_pkgs An optional character string for which packages are
+#' required.
+#' @param seed A single integer used to control randomness when models are
+#' (re)fit.
+#' @return A list object with the options given by the user.
+#' @export
+control_conformal_infer <-
+  function(method = "search", trial_points = 100, var_multiplier = 10,
+           max_iter = 100, tolerance = .Machine$double.eps^0.25, progress = FALSE,
+           required_pkgs = character(0), seed = NULL) {
```

Comment:
Can we just set a default `seed = sample.int(10^5, 1)` and avoid the `is.null` conditional?